rir/src/ir/Compiler.cpp

```diff
@@ -145,6 +145,143 @@ Code* compilePromise(CompilerContext& ctx, SEXP exp);
 void compileExpr(CompilerContext& ctx, SEXP exp);
 void compileCall(CompilerContext& ctx, SEXP ast, SEXP fun, SEXP args);
 
+void compileWhile(CompilerContext& ctx, std::function<void()> compileCond,
+                  std::function<void()> compileBody) {
+    CodeStream& cs = ctx.cs();
+
+    BC::Label loopBranch = cs.mkLabel();
+    BC::Label nextBranch = cs.mkLabel();
+
+    ctx.pushLoop(loopBranch, nextBranch);
+
+    unsigned beginLoopPos = cs.currentPos();
+
+    cs << BC::beginloop(nextBranch) << loopBranch;
+
+    compileCond();
+    cs << BC::brfalse(nextBranch);
+
+    compileBody();
+    cs << BC::pop() << BC::br(loopBranch) << nextBranch;
+
+    if (ctx.loopNeedsContext()) {
+        cs << BC::endloop();
+    } else {
+        cs.remove(beginLoopPos);
+    }
+
+    cs << BC::push(R_NilValue) << BC::invisible();
+
+    ctx.popLoop();
+}
+
+bool compileSimpleFor(CompilerContext& ctx, SEXP sym, SEXP seq, SEXP body) {
+    Match(seq) {
+        Case(LANGSXP, fun, argsSexp) {
+            RList args(argsSexp);
+            if (fun != symbol::Colon || args.length() != 2) {
+                return false;
+            }
+
+            SEXP start = args[0];
+            SEXP end = args[1];
+            if (TYPEOF(start) != INTSXP && TYPEOF(start) != LGLSXP &&
+                (TYPEOF(start) != REALSXP || XLENGTH(start) == 0 ||
+                 *REAL(start) != (int)*REAL(start))) {
+                return false;
+            }
```

Comment:
so start has to be a constant, whereas end can be evaluated, right? what's the reason for this choice?

Reply:
If `start` is an (actual, not in R types) integer than the list will guaranteed either only contain integers (even if `end` is a real), or raise an error (if `end` isn't a number). If `start` isn't an integer than the list could contain non-integers, and then we'd have to `inc` and `dec` reals (which would be possible but more complicated). The best way to check if `start` is an integer is by looking if its an integer literal. Ultimately, most programs will have `for (i in <constant>:n)`, so its easier to only handle those cases.

Reply:
ok, I got it. actual integer literals are not very common, since you have to explicitly write the trailing `L`. but maybe you could also do this to real literals, which happen to be an integer?

Reply:
Sorry I didn't use the right term. This checks if `start` is a literal which is a "true" integer, so if `start` is an R-type real literal it can still compile the simple range.

Reply:
right, now I actually read the thing.